File: fase_2/building_roads.py
from collections import defaultdict
import sys

sys.setrecursionlimit(2**31-1)

def dfs(s):
    if vis[s]:
        return
    vis[s] = True
    for i in adj[s]:
        dfs(i)

n, m = map(int, input().split())

adj = defaultdict(list)
vis = [False]* (n+1)

for _ in range(m):
    u, v = map(int, input().split())
    adj[u].append(v)
    adj[v].append(u)
v = []
for i in range(1,n+1):
    if not vis[i]:
        v.append(i)
        dfs(i)

print(len(v)-1)
for i in range(1, len(v)):
    print(v[i], v[i-1])

File: fase_2/building_roads.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long int ll;
const int MAX = 1e5+10;
#define endl '\n'

vector<vector<int>> adj(MAX);
vector<bool> vis(MAX,false);

void dfs(int num){
    if (vis[num]) return;
    vis[num]=true;

    for (auto x : adj[num]){
        dfs(x);
    }

}



int main(){
    int n,m; cin >> n>>m;

    for (int i=0; i<m;i++){
        int x, y; cin>> x >> y; x--; y--;
        adj[x].push_back(y);
        adj[y].push_back(x);
    }

    vector<int> v;

    for (int i=0; i<n; i++){
        if (!vis[i]){
            v.push_back(i);
            dfs(i);
        }
    }

    int len = v.size()-1;
    cout << len << endl;

    for (int i=1; i<=len; i++){
        cout << v[i-1]+1<< " "<< v[i]+1 << endl;
    }
}